File: javascript/satellite/satellite.js
//
// This is only a SKELETON file for the 'Satellite' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

/**
 * Reconstruct a binary tree with the information from the given pre-order and 
 * in-order traversals.
 * @param {Array} preorder - tree as array of values in pre-order traversal order
 * @param {Array} inorder - tree as array of values in in-order traversal order
 * @returns {Object} - tree node object with values "value" a string, "left", and "right". Left and right are 
 * empty objects for leaf nodes ({}), or another tree node object
 */
export const treeFromTraversals = (preorder, inorder) => {

  if (preorder.length !== inorder.length)
    throw new Error('traversals must have the same length')

  if ((new Set(preorder).size !== preorder.length) ||
      (new Set(inorder).size !== inorder.length))
    throw new Error('traversals must contain unique items')
  
  if ([...preorder].sort().join('') !== [...inorder].sort().join(''))
    throw new Error('traversals must have the same elements')
  
  if (preorder.length === 0)
    return {}

  return traverse([...preorder], inorder, 0, preorder.length - 1)
};

function traverse(preorder, inorder, start, stop)
{
  let root = preorder.shift();
  if (typeof root === 'undefined')
    return {}
  let rootIndex = inorder.findIndex((value, index) => value === root && index >= start && index <= stop)
  let left = {}
  if (rootIndex - start > 0)
  {
    left = traverse(preorder, inorder, start, rootIndex - 1)
  }
  let right = {}
  if (stop - rootIndex > 0)
  {
    right = traverse(preorder, inorder, rootIndex + 1, stop)
  }

  return { value: root, left: left, right: right}
}

File: csharp/satellite/Satellite.cs
using System.Reflection.Metadata.Ecma335;

public record Tree(char Value, Tree? Left, Tree? Right);

public static class Satellite
{
    /// <summary>
    /// Reconstruct a binary tree with the information from the given pre-order and 
    /// in-order traversals.
    /// </summary>
    /// <param name="preOrder">tree as array of values in pre-order traversal order</param>
    /// <param name="inOrder">tree as array of values in in-order traversal order</param>
    /// <returns>Tree? object, null if an empty tree, otherwise filled with reconstructed tree</returns>
    /// <exception cref="ArgumentException">Thrown if argments are not the same length, don't have 
    /// unique items (no repeats) or the same elements in both</exception>
    public static Tree? TreeFromTraversals(char[] preOrder, char[] inOrder)
    {
        if (preOrder.Length != inOrder.Length)
            throw new ArgumentException("traversals must have the same length");

        if ((new HashSet<char>(preOrder).Count != preOrder.Length) ||
            (new HashSet<char>(inOrder).Count != inOrder.Length))
            throw new ArgumentException("traversals must contain unique items");
  
        String preText = String.Join("", preOrder.Select(c => c.ToString()).OrderBy(key => key));
        String inText = String.Join("", inOrder.Select(c => c.ToString()).OrderBy(key => key));
        if (inText != preText)
            throw new ArgumentException("traversals must have the same elements");

        if (preText.Length == 0)
            return null;

        Queue<char> preorderQueue = new(preOrder);
        return traverse(preorderQueue, inOrder, 0, inOrder.Length - 1);
    }

    private static Tree? traverse(Queue<char> preorder, char[] inorder, int start, int stop)
    {
        if (!preorder.TryDequeue(out char root))
            return null;

        int rootIndex = inorder
                .Index()
                .Where(pair => pair.Item == root && pair.Index >= start && pair.Index <= stop)
                .Select(pair => pair.Index).First();

        Tree? left = null;
        Tree? right = null;        
  
        if (rootIndex - start > 0)
        {
            left = traverse(preorder, inorder, start, rootIndex - 1);
        }

        if (stop - rootIndex > 0)
        {
            right = traverse(preorder, inorder, rootIndex + 1, stop);
        }

        return new Tree(root, left, right);
    }
}
